test/connector/tcp/mssql/test

```diff
@@ -1,17 +1,133 @@
-#!/bin/bash
+#!/bin/bash -e
 
-export SECRETLESS_HOST=secretless
+readonly db_host_tls="mssql"
+readonly db_host_no_tls="mssql" # TODO: configure a non-ssl container?
+readonly db_port=1433
+readonly db_user="sa"
+readonly db_password="yourStrong()Password"
+readonly secretless_port=2223
+secretless_host="secretless"
+benchmark=false
+
+help_and_exit() {
+  local retval=${1:-1}
+  cat <<EOF
+This script tests access to an MSSQL server, first accessing the
+server directly (to establish a testing baseline), and then
+via a secretless broker.
+
+This script detects if you're in "dev" mode by the existence of
+the secretless-dev container.  That container will exist only if
+the development process was started using the "./dev" command.
+
+Options:
+  -b runs benchmarks
+EOF
+  exit "$retval"
+}
+
+while getopts :b opt; do
+    case $opt in
+        b) benchmark=true;;
+       \?) echo "Unknown option -$OPTARG"; help_and_exit 1;;
+    esac
+done
+
+assert_contains() {
+  # assert_contains <string> <substring>
+  local string=$1
+  local substring=$2
+
+  if [[ "${string}" =~ ${substring} ]]; then
+      echo "."
+      return
+  fi
+
+  echo ""
+  echo "FAIL:"
+  echo ""
+  echo "The string:"
+  echo ""
+  echo "    ${string}"
+  echo ""
+  echo "Did not contain expected substring:"
+  echo ""
+  echo "    ${substring}"
+  echo ""
+
+  exit 1
+}
+
+query_sql_service() {
+  # query_sql_service <server> <port> <user> <password>
+  server=$1
+  port=$2
+  user=$3
+  pw=$4
+
+  docker_cmd="docker-compose run --rm --no-deps"
+  sql_cmd=(sqlcmd -S "$server,$port" -U "$user" -P "$pw" -Q "SELECT 1+1")
+
+  # NOTE: We need to 2>&1 because we need to assert on sqlcmd's error messages
+  # and we need to "|| true" because we don't want the script to exit when this
+  # happens
+  $docker_cmd test "${sql_cmd[@]}" 2>&1 || true
+}
+
+# Automatically detect if we're devmode based on the existence
+# of the secretless-dev container.  We assume that you started
+# your workflow using `./dev` if you are developing, and this
+# command will use the secretless-dev container.
 if [[ ! -z $(docker-compose ps -q secretless-dev) ]]; then
-  export SECRETLESS_HOST=secretless-dev
+  secretless_host="secretless-dev"
```

Comment:
comment on existing code, but since we're here we should change:
```sh
if [[ ! -z $(docker-compose ps -q secretless-dev) ]]; then
```
to (untested):
```sh
if docker-compose ps -q secretless-dev 2>&1; then
```

Reply:
That doesn't seem to work. What I'm seeing when I try this is that conditional that you're suggesting looks at the `exit value` of the `docker-compose ps ...` command. In this case, the `docker-compose ps ...` command returns a success exit value of 0 whether the `secretless-dev` container is running or not. So we need to look at stdout from this command, i.e. use command substitution. Re. the stderr, I *think* we want to ignore stderr in this case, and base our decision solely on whether or not there's stdout for the command. One minor improvement, though, would be to use `-n` instead of `! -z`:
```
if [[ -n $(docker-compose ps -q secretless-dev) ]]; then
```

Reply:
From what I'm seeing the exit code should work:

```sh
~/cyberark/secretless-broker/src/test/connector/tcp/mssql (master) $ dc ps
         Name                       Command                State             Ports         
-------------------------------------------------------------------------------------------
mssql_mssql_1            /opt/mssql/bin/nonroot_msg ...   Up         0.0.0.0:1433->1433/tcp
mssql_secretless-dev_1   ./bin/reflex                     Up         0.0.0.0:2223->2223/tcp
mssql_secretless_1       /usr/local/bin/secretless- ...   Exit 128                         
mssql_test_1             sleep 999d                       Up                               
~/cyberark/secretless-broker/src/test/connector/tcp/mssql (master) $ dc ps -q secretless-dev; echo $?
8ca8cacd29772cb5f294db45f6d4227ecb4eb88a8eeff40a846ab00fec26d373
0
~/cyberark/secretless-broker/src/test/connector/tcp/mssql (master) $ dc ps -q xxxxxx; echo $?
ERROR: No such service: xxxxxx
1
```

Reply:
Oh also, since quiet still displays the IDs we want:
```sh
if docker-compose ps --quiet secretless-dev 1>/dev/null 2>&1; then
```

Reply:
Nevermind, you're correct.  Ignore my above comments.  It still succeeds if the service exists but it not running.